sympy/matrices/normalforms.py

```diff
@@ -0,0 +1,115 @@
+from sympy.matrices import *
+
+'''
+Functions returning normal forms of matrices
+
+'''
+def smith_normal_form(m, domain = None):
+    invs = smith_normal_invariants(m)
+    smf = diag(*invs)
+    n = len(invs)
+    if m.rows > n:
+        smf = smf.row_insert(m.rows, zeros(m.rows-n, m.cols))
+    elif m.cols > n:
+        smf = smf.col_insert(m.cols, zeros(m.rows, m.cols-n))
+    return smf
+
+def smith_normal_invariants(m, domain = None):
+    '''
+    Return the tuple of abelian invariants for a matrix `m`
+    (as in the Smith-Normal form)
+
+    '''
+    if not domain and not (hasattr(m, "ring") and m.ring.is_PID):
+        raise ValueError(
+            "The matrix entries must be over a principal ideal domain")
+    elif not domain:
```

Comment:
This is a bit confusing. The two conditionals here are mixing things up. Let's try to break things down. 
1. Check the `domain` argument.
2. If it is `None`, we decide the domain based on the ring of the matrix entries.
3. Check if the entries of the matrix are defined over a PID

Here is a sample implementation which might help:
```python
if not domain:
    if not (hasattr(m, "ring") and m.ring.is_PID):
        raise ValueError()
    else:
        domain = m.ring